### ISSUE TITLE
How are multi-metric scores are represented in CWMNX

### ISSUE THREAD

**Issue by joeberkovitz** [w3c/mnx#55]
The `<global>` element is at odds with the idea of scores whose parts have different meters, and thus do not have the same number of measures.

Multi-metric scores can be supported by ignoring `<global>`, and placing notations that would otherwise be global into specific parts. This is one possible solution but there may be others.

**Comment by cecilios** [comment]
Apart from `<global>` I think that for now current draft is valid for dealing with multi-metric scores. The problem with `<global>` could be solved as follows: 

- `<global>` will be not mandatory.

- Including all measures in `<global>` is not mandatory: `<global>` should not be a 'grid' for the score but just a place for common directions. Nevertheless, nothing prevents to used it as a grid if desired but I would promote this idea from the standard.

- Do not include empty `<measures>` in `<global>`. Include only those measures that have common content. For this, it is mandatory that the included measures have a reference to point to the affected parts measures.

- Apart from the 'number' attribute in `<measure>` define a new optional attribute, 'gref' (global ref.), that will behave as rehearsal marks do: as a common reference point for measures in all parts.

- For `<measure>` elements included in `<global>` it will be mandatory that they include either the 'number' attribute or the 'gref' attribute:
	- For regular scores with time signature, any of them is suitable.
	- For multi-metric scores, always 'gref' must be used.

I attach an example.

```<?xml version="1.0" encoding="UTF-8"?>
<!-- Note: page-layout-related information and implied stem directions have been removed from this example -->
<mnx>
  <head>
    <title>Multi-metric example</title>
    <style-selector rule="score" style="staff-distance: 80px; staff-line-spacing: 18px; font-family: Times New Roman; font-size: 10px; smufl-font: Maestro;"/>
    <style-class name="tempo-indication" style="font-weight: bold;"/>
  </head>

  <score>
    <cwmnx profile="standard">
      <!-- Global directions -->
      <global>
        <measure gref="1">
          <directions>
            <tempo bpm="60" value="/4"/>
            <clef line="2" sign="G" staff="1"/>
            <key fifths="1" mode="major"/>
            <instruction class="tempo-indication">Andantino</instruction>
          </directions>
        </measure>
        <measure gref="2"/>
          <directions>
            <time signature="3/4"/>
          </directions>
        </measure>
      </global>

      <!-- Part 1 =======================================================-->
      <part style="measure-numbering: none;">
        <part-name>Violin</part-name>
        <instrument-sound>violin</instrument-sound>
        <measure number="1" gref="1">
          <directions>
            <time signature="3/4"/>
          </directions>
          <sequence staff="1">
            <event value="/4">
              <note pitch="G3"/>
              <note pitch="D4"/>
            </event>
            <event value="/8">
              <rest/>
            </event>
            <event value="/8">
              <note pitch="G5"/>
            </event>
            <beamed>
              <event value="/16">
                <note pitch="G5"/>
              </event>
              <event value="/16">
                <note pitch="F5"/>
              </event>
              <event value="/8">
                <note pitch="G5"/>
              </event>
            </beamed>
          </sequence>
        </measure>

        <measure number="2">
          <sequence staff="1">
            <event measure="yes">
              <rest/>
            </event>
          </sequence>
        </measure>

        <measure number="3">
          <sequence staff="1">
            <event value="/2">
              <note pitch="G3"/>
              <note pitch="D4"/>
            </event>
            <event value="/4">
              <note pitch="G3"/>
              <note pitch="D4"/>
            </event>
          </sequence>
        </measure>

        <measure number="4">
          <sequence staff="1">
            <event value="/4">
              <note pitch="A4"/>
              <note pitch="E5"/>
            </event>
            <event value="/4">
              <rest/>
            </event>
            <event value="/4">
              <note pitch="D4"/>
              <note pitch="G4"/>
              <note pitch="F5"/>
            </event>
          </sequence>
        </measure>

        <measure number="5" gref="2">
          <sequence staff="1">
            <event measure="yes">
              <rest/>
            </event>
          </sequence>
        </measure>

      </part>

      <!-- Part 2 =========================================================-->
      <part style="measure-numbering: none;">
        <part-name>Violin</part-name>
        <instrument-sound>violin</instrument-sound>
        <measure number="1" gref="1">
          <directions>
            <time signature="2/4"/>
          </directions>
          <sequence staff="1">
            <event value="/4">
              <note pitch="G4"/>
            </event>
            <beamed>
              <event value="/8">
                <note pitch="D5"/>
              </event>
              <event value="/8">
                <note pitch="D5"/>
              </event>
            </beamed>
          </sequence>
        </measure>

        <measure number="2">
          <sequence staff="1">
            <beamed>
              <event value="/8">
                <note pitch="B5"/>
              </event>
              <event value="/16">
                <note pitch="A5"/>
              </event>
              <event value="/16">
                <note pitch="G5"/>
              </event>
            </beamed>
            <beamed>
              <event value="/8">
                <note pitch="G5"/>
              </event>
              <event value="/8">
                <note pitch="G5"/>
              </event>
            </beamed>
          </sequence>
        </measure>

        <measure number="3">
          <sequence staff="1">
            <beamed>
              <event value="/8">
                <note pitch="E5"/>
              </event>
              <event value="/16">
                <note pitch="D5"/>
              </event>
              <event value="/16">
                <note pitch="C5"/>
              </event>
            </beamed>
            <beamed>
              <event value="/8">
                <note pitch="E5"/>
              </event>
              <event value="/8">
                <note pitch="E5"/>
              </event>
            </beamed>
          </sequence>
        </measure>

        <measure number="4">
          <sequence staff="1">
            <beamed>
              <event value="/8">
                <note pitch="D5"/>
              </event>
              <event value="/8">
                <note pitch="G5"/>
              </event>
              <event value="/8">
                <note pitch="G5"/>
              </event>
              <event value="/8">
                <note pitch="G5"/>
              </event>
            </beamed>
            <barline>
          </sequence>
        </measure>

        <measure number="5">
          <sequence staff="1">
            <beamed>
              <event value="/8">
                <note pitch="B5"/>
              </event>
              <event value="/8">
                <note pitch="G5"/>
              </event>
            </beamed>
            <beamed>
              <event value="/16">
                <note pitch="E5"/>
              </event>
              <event value="/16">
                <note pitch="C5"/>
              </event>
              <event value="/16">
                <note pitch="E5"/>
              </event>
              <event value="/16">
                <note pitch="C5"/>
              </event>
            </beamed>
          </sequence>
        </measure>

        <measure number="6">
          <sequence staff="1">
            <beamed>
              <event value="/16">
                <note pitch="B4"/>
              </event>
              <event value="/16">
                <note pitch="D5"/>
              </event>
              <event value="/16">
                <note pitch="B4"/>
              </event>
              <event value="/16">
                <note pitch="D5"/>
              </event>
            </beamed>
            <beamed>
              <event value="/8">
                <note pitch="C5"/>
              </event>
              <event value="/16">
                <note pitch="B4"/>
              </event>
              <event value="/16">
                <note pitch="A4"/>
              </event>
            </beamed>
          </sequence>
        </measure>

        <measure number="7" gref="2">
          <sequence staff="1">
            <event measure="yes">
              <rest/>
            </event>
          </sequence>
        </measure>

      </part>
    </cwmnx>
  </score>
</mnx>
```

**Comment by cecilios** [comment]
Thinking more about the multi-metric scores problem, I now understand the rationale for a lot of empty `<measure>` elements in `<global>` in your examples.

So to avoid more mistakes, I would appreciate some clarifications about the rationale for the changes introduced by the `<global>` element (previously `<system>` element) in the latest draft.

The initial argument, when the `<system>` element was created, was to notate elements that due to a global semantic constrain, if they were notated separately in each part, this notation must agree across all parts (e.g. tempo indications, rehearsal marks, or system and page breaks are global to all parts). In other cases (time signatures, for example) while such agreement is not necessarily the case, it seems desirable to allow notate this agreement in those cases in which the agreement is a requirement.

In current draft, as the `<global>` element is defined ("The global element represents a set of measures, each of which provides content that is shared by all parts within the score. Each measure element within global supplies the shared content for all other measure elements which share the same measure number."), we have lost the idea of notating a global constrain. Why this change? As it is now defined, it transmits the idea of an artifact for optimizations, to avoid to encode multiple times common content.

Nevertheless, I think that the root of the problem with multi-metric scores is in the definition of the 'standard' profile, as it restricts this profile to scores with equal number of measures in all parts. I'm not sure if the 'standard' profile should be so restrictive: Mozart will no longer be CWMN ;-)

Jokes apart, a solution for multi-metric could be to define another cwmnx profile, e.g. multi-metrics. It would allow better control for these rare cases as their constrains are different. But the 'multi-metrics' profile should be developed in parallel to the 'standard' profile, so that do not transmit the idea that this profile is something different from the 'standard' and to minimize the risks that this profile will not be supported by software supporting the standard profile. For this, perhaps the best approach would be to make the 'standard' profile description applicable to 'multi-metrics` except when explicitly described.

**Comment by joeberkovitz** [comment]
I think some clarification of the spec is needed. The reasoning for `<global>` is exactly the same as `<system>`, this was intended as just a name change and nothing else. `<global>` is not about optimizing the encoding: it's a way of describing measure content (e.g. tempo indications) that are true global constraints, or which transcend the notion of "part" and are thus shown in the score regardless of which subset of parts are visible in a rendering. 

You're on the same track as me: my thought is to have another cwmnx profile for multi-metric scores. My hope is that it would be widely supported. However, let's be honest about the world out there: not all notation software does support multi-metric scores, and they introduce a great deal of additional implementation burden that some may not want to take on. So I personally think there is value in isolating a simpler standard use case, even if it does fail to support a tiny number of works by Western-canon composers.

**Comment by cecilios** [comment]
I fully agree. And that is the reason for my suggestion of making the 'standard' profile description applicable to 'multi-metrics` except when explicitly described. If the exceptions are not many, perhaps implementers will more likely support the new profile.

**Comment by joeberkovitz** [comment]
@cecilios As we've cleared up the problem with `<global>` and multi-metric by fixing #2, I am wondering if we should close this issue also. The question of the definition of the `standard`  profile is still open but I think this issue was more about how we'd represent multi-metric scores, not what profile they belong to.

**Comment by cecilios** [comment]
Sorry for the delay in answering. I've been on vacation for two weeks without internet and have returned this morning.

I agree that the issue of multi-metric representation is nearly ready to be closed. But before closing I think it is necessary to fix a couple of small issues.

One is the example 9. Perhaps I've missed something but the measure element has attributes index, number and barline but no attribute 'sequence', and example 9 refers to attribute 'sequence' that needs clarification.

The second issue is defining a profile in which using the global element restricted to parts is valid. PR #81 just says that is not part of the standard profile but does not say which profile or profiles accept it.

For me the profile for multi-metric scores is less important than:
1. Agreeing on the name and scope of the standard profile. I would suggest to open an issue for getting feedback from the group on the name and scope of the standard profile. The name 'standard' means nothing if we have not agreed on what is standard and the name 'standard' profile could miss-guide implementors on the assumption that by implementing the  'standard' profile they will cover all CWMN music.

2. If the MNX-Common standard is going to be split in profiles, I suggest that the MNX document should include a section explaining all the profiles needed for covering CWMN music and clear descriptions about what types of scores will be left uncovered by each profile and not just a list of technical restrictions.

**Comment by notator** [comment]
@cecilios I think we need some clarification from the chair about "If the MNX-Common standard is going to be split in profiles..."
As I understand https://github.com/w3c/mnx/issues/74#issuecomment-374265850, MNX-Common is the name of the profile called "standard" in the Draft Spec, not the name of the set of all CWMN profiles. If I'm right, then future profiles would be named using the following pattern: MNX-Baroque, MNX-GuitarTab etc. (not MNX-Common-Baroque, MNX-Common-GuitarTab etc.)

**Comment by cecilios** [comment]
@bhamblock In PR #81 if you click on link 'preview' in first post, you will arrive to the updated MNX Draft Specification. Go to 'The global element' section (5.2.3.2). I'm referring to Example 9 in that section.

@notator Perhaps I misunderstood the name changes. My understanding is that MNX-Common replaces old name CWMNX, that included among other future profiles, the 'standard' profile. And that MNX-Generic is was named GMNX.

**Comment by joeberkovitz** [comment]
@cecilios: answering your question above: issue #68 (which you opened) is already a marker recording the need to define the standard profile with some care. So I don't think we need a new issue: it's agreed that we need a careful definition of profiles. So far we have deferred this discussion, because we want to make sure that at the very least there is one profile whose scope is roughly equivalent to MusicXML. But this does not exclude others.

Thanks for catching the broken references to the `sequence` attribute - a typo that is fixed by #91.

@cecilios Your understanding of profiles is correct: MNX-Common is just a renaming of CWMNX, with no change in the definition of profiles. Thus, there **may** be multiple profiles of MNX-Common in the future, which is what #65 is about. @notator the `profile` attribute of an MNX-Common score identifies its profile. We have not changed this in any way. No identification of profiles with musical idioms  _within CWMN_ has been agreed on, and that question is off topic for this issue which is about multi-metric scores.

**Comment by notator** [comment]
@joeberkovitz Thanks for the clarification.

**Comment by cecilios** [comment]
@joeberkovitz Thanks for the clarifications and for pointing to #68. I wasn't aware of its existence :)

Then, I agree that this issue can be close.

**Comment by clnoel** [comment]
I don't think this issue can be closed until we have specified another profile in the spec for multi-metric, even if it is just a place holder. Such a profile should allow more than one "global" element, which the standard profile does NOT allow.

**Comment by joeberkovitz** [comment]
@clnoel agreed, I think we need to come back to #68 and refine it first. I believe that we should step away from the "standard" dilemma and redefine profiles as a set of completely independent constraints, each of which is perfectly well-defined and objective (for instance, "score is not multi-metric" would be one such constraint). Then, we can argue later about what set of constraints, if any, constitute "standard".  (That's my personal opinion, I'm not speaking for all the co-chairs here.)

**Comment by cecilios** [comment]
@joeberkovitz  I agree with your proposal

**Comment by bhamblok** [comment]
Sorry if this is the wrong issue to write this comment.
@adrianholovaty added a "Time signatures" section on the "By Example"-page. At first sight this is looking good, except when we will be handling multi-metric scores.
In my vision, every visual element on a score, should be represented by an xml-tag... So should be a time-signature, for each individual part. So I think the "time"-tag does not belong into the global section, but rather in the "directions"-section for each individual part ... or maybe even for each individual staff (=section) (just like a clef).
MNX should still become the digital representation of the visual score. Just like SVG represents images, it does not state anything about the contents of the image, it only states the shapes used to create an image.

**Comment by bhamblok** [comment]
Sorry if this is the wrong issue to write this comment.
@adrianholovaty added a "Time signatures" section on the "By Example"-page. At first sight this is looking good, except when we will be handling multi-metric scores.

In my vision, every visual element on a score, should be represented by an xml-tag... So should be a time-signature, for each individual part. So I think the "time"-tag does not belong into the global section, but rather in the "directions"-section for each individual part ... or maybe even for each individual staff (=section) (just like the clef element).
Why on a section (=staff) level you might ask? Imagine a simple (but multi-metric) piano-score, having 2 staves in one part, each with its own time signature...

My vision in short: MNX should still become the digital representation of the **visual** score. (For a digital representation of an **audible** score we have MIDI :-)). And just like SVG is representing images, where SVG does not state anything about the contents of the image, but only describes the shapes used to create an image; MNX should describe every element visually on the score. With the extra semantics of the MNX-format, which also describes the context of all the elements, we will have the extra beneficial information to be able to describe the audible content as well.

**Comment by clnoel** [comment]
@bhamblok While it is important that the MNX specification be able to reproduce the original visual score, a lot of the point of what we are doing here is to make the MNX document flexible enough that new layouts (page size, part extractions, etc.) and audio extractions are possible. Hence why we had such a huge debate about how to represent octave shifts and transposing instruments. As such, the global element should be considered to be the place where multi-part notations go. Every notation in the global element is implicitly replicated to all parts. 

Therefore, in single-metric pieces, the time element goes in the global element, and the visual display of that is replicated across all parts.

In multi-metric pieces, if I'm remembering correctly, the current concept is that we have more than one "global" element (possibly renamed), with specifications for which parts belong to which global.
